feat: add persistent comment drafts

nodogx opened the pull request:
# Description

This Pull Request introduces a new feature that enables comment drafting in the task comment composer. User (all users including tutors) can now have their input saved per task using 'localStorage'. This prevents accidental data loses during task switches or page reloads. This can be also used to draft up something and send it at a later stage. (Something like quick notes) 



- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

This has been tested manually, by drafting comments with the reply to behavior. It has also been tested with uploading files to the comment section etc, to ensure everything work as intended.

This has also been tested with the Browser's Local Storage to ensure how things are stored per character, and how it deletes and stores.

## Video demonstration of how this new feature works:

https://github.com/user-attachments/assets/7eab18b8-cd96-480c-be75-c2558a8443af




## Testing Checklist:

- [x] Tested in latest Chrome
- [x] Tested in latest Safari
- [x] Tested in latest Firefox

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have requested a review from @macite and @jakerenzella on the Pull Request

s223749059 committed:
refactor: migrate the confirmation-model component

s223749059 committed:
refactor: migrate the confirmation-model component - updates

b0ink committed:
refactor: create initial files for migration of comments-modal

b0ink committed:
refactor: create service file for migration of comments-modal

b0ink committed:
refactor: migrate comments-modal

b0ink committed:
refactor: remove old comments-modal files

b0ink committed:
fix: match comments-modals's original max width

b0ink committed:
refactor: enforce type safety for data passed to comments-modal component

b0ink committed:
fix: allow oversized images to be scrollable

b0ink committed:
refactor: replace sass with tailwind css

b0ink committed:
refactor: only reveal scrollbar on oversized images

b0ink committed:
feat: persistent pdf zoom level

b0ink committed:
refactor: replace bootstrap html with material components

b0ink committed:
refactor: improve modal size

b0ink committed:
fix: remove duplicate alert

b0ink committed:
refactor: align caption with modal title

JackSCarroll committed:
refactor: add button on frontend for future jplag report viewer

JackSCarroll committed:
feat: add jplag report download button to student submission tutor teacher footer view

JackSCarroll committed:
feat: jplag report download button in task explorer view

JackSCarroll committed:
chore: commit edited files

b0ink committed:
Merge remote-tracking branch 'doubtfirelms/8.0.x' into migrate/comments-modal

b0ink committed:
refactor: replace iframe with native pdf-viewer

b0ink committed:
Merge remote-tracking branch 'doubtfirelms/8.0.x' into migrate/confirmation-modal

b0ink committed:
feat: add startpage property to auto scroll to on pdf render

b0ink committed:
feat: init pdf viewer on page 2 for student submissions

b0ink committed:
Merge remote-tracking branch 'doubtfirelms/8.0.x' into feat/persistent-pdf-zoom

JackSCarroll committed:
refactor: add features for jplag report download

JackSCarroll committed:
chore: comment out staff task list jplag report download button until fix

satikaj committed:
feat: add editor for task feedback templates

satikaj committed:
feat: combine outcome and template editor

satikaj committed:
feat: add field to connect outcomes

satikaj committed:
feat: add feedback template picker to comment composer

satikaj committed:
feat: make template editor reusable across contexts

satikaj committed:
fix: remove selected outcomes from dropdown

macite committed:
feat: switch to html5 mode

satikaj committed:
feat: modify fields and operate on learning outcomes

satikaj committed:
refactor: remove old outcome components from view

macite committed:
feat: add ability to collect d2l org and grade object ids

macite committed:
fix: ensure standard error propagate to the ui

macite committed:
chore: add ability for unit to detect changes

macite committed:
feat: add ability to trigger d2l grade transfer

macite committed:
fix: report better errors during file downloads

satikaj committed:
feat: add connected learning outcomes

satikaj committed:
refactor: change type name to Template

satikaj committed:
fix: save to proper cache and clean up

satikaj committed:
feat: add glo editor

satikaj committed:
fix: save connected outcomes

satikaj committed:
refactor: cleanup feedback editor html

satikaj committed:
fix: tlo csv url

macite committed:
chore: allow development via external tunnel without port

macite committed:
feat: d2l integration checks settings

- checks with weighted grading
- provides a link to open D2L unit edit for weight

satikaj committed:
refactor: cleanup feedback picker

macite committed:
feat: check if d2l results available

macite committed:
feat: add success page for oauth callback

satikaj committed:
fix: save linked outcomes upon selection

satikaj committed:
fix: check difference before updating linked outcomes

satikaj committed:
feat: edit feedback chips

satikaj committed:
fix: template editor not showing if new

satikaj committed:
fix: reset template table for new outcome

satikaj committed:
fix: hide template editor on outcome switch

satikaj committed:
fix: upon save update entity in frontend

satikaj committed:
fix: only get feedback chips if staff

satikaj committed:
fix: refresh context when changing task defs

satikaj committed:
fix: refine feedback editor

satikaj committed:
refactor: remove nested feedback csv and add outcome csv modal

satikaj committed:
refactor: remove unused feedback picker sections

satikaj committed:
feat: show chips on picker

satikaj committed:
feat: add comment text to comment area

macite committed:
Merge pull request #909 from macite/new/d2lintegration

New/d2lintegration - Introduce D2L integration for grade transfer

satikaj committed:
fix: hide picker on task change

satikaj committed:
feat: fetch feedback chips when selecting task in unit editor

satikaj committed:
fix: show proper selected outcome or template

macite committed:
chore: update run all and remove watch for angularjs changes

macite committed:
chore(release): 8.0.30

macite committed:
fix: ensure test attempt work for staff students

Where the staff member is the student.
Also ensures the current user is in the cache when they login from cookie.

satikaj committed:
fix: organise outcome and feedback csv functionality

satikaj committed:
fix: remove glo outcome specific csv buttons

satikaj committed:
fix: check for unit context for template csv name

satikaj committed:
fix: adjust ui for nested csv modal

satikaj committed:
fix: feedback picker search

satikaj committed:
fix: hide feedback group comment text

coskun-kilinc committed:
feat(feedback-template): add hierarchical sorting for feedback templates

- Added `mat-sort-header` to `chipText` and `description` table headers
  to enable sorting.
- Implemented custom sorting logic in `sortTemplateData()` to preserve
  parent-child hierarchy.
- Standardised sorting logic for `chipText`, `description`, and
  `taskStatus` fields.
- Refactored `sortTemplateData()` to group and sort child templates
  within their respective parents.
- Applied consistent formatting to imports and component decorators.

coskun-kilinc committed:
feat(feedback-template): merge upstream changes

coskun-kilinc committed:
fix(feedback-template): ensure sorting defaults are applied

- Applied default sorting for feedback templates when when fetched.
- Ensured `sortTemplateData` uses 'chipText' as the default column and
  'asc' as the default direction.

coskun-kilinc committed:
feat(feedback-template): improve sorting logic for feedback templates

- Refactored `sortTemplateData` to implement a dynamic sorting function
  based on a selected property.
- Introduced a depth-based hierarchical sorting mechanism to better
  handle grouped feedback templates.
- Added step-by-step logging to debug sorting logic and ensure correct
  parent-child relationships.
- Assigned sequential order numbers and adjusted them based on hierarchy
  depth.
- Improved the final sorting step to maintain a consistent and logical
  structure.
- Replaced hard-coded sorting conditions with a generic property-based
  comparator.

macite committed:
fix: serve index.html for unknown paths

This should ensure that the new html5 urls work correctly.

macite committed:
chore(release): 8.0.31

macite committed:
fix: correct issue fetching user from cookie

macite committed:
chore(release): 8.0.32

macite committed:
fix: nginx conf redirect for spa

- Add directory to direct to index.html
- Add root index.html otherwise

macite committed:
chore(release): 8.0.33

satikaj committed:
refactor: enhance feedback picker ui

satikaj committed:
fix: show picker above composer and below comments

satikaj committed:
fix: focus on composer input after selecting chip

satikaj committed:
fix: display task status in coloured chip in feedback editor

satikaj committed:
fix: display feedback picker chips in color

satikaj committed:
refactor: add feedback chip icons

satikaj committed:
fix: hide summary instead of comment text for groups

satikaj committed:
fix: use icon instead of color for template task status

satikaj committed:
refactor: add outcome heading to picker

satikaj committed:
fix: display nested feedback groups in picker

satikaj committed:
feat: suggest task status according to feedback

satikaj committed:
Merge branch 'add-tutor-feedback' into add-tutor-feedback

satikaj committed:
Merge pull request #1 from coskun-kilinc/add-tutor-feedback

Add hierarchical sorting for feedback templates

satikaj committed:
fix: comment console logs

coskun-kilinc committed:
fix(feedback-template-editor): improve feedback template sorting logic

- Separated feedback groups and non-group templates early for
  efficiency.
- Improved depth calculation by explicitly iterating over groups.
- Optimised hierarchical ordering by refining the depth-based
  multiplier.
- Removed redundant logging and consolidated essential debug
  information.
- Ensured computed hierarchical order is consistently applied across
  groups and templates.

satikaj committed:
Merge branch 'add-tutor-feedback' into add-tutor-feedback

satikaj committed:
Merge pull request #2 from coskun-kilinc/add-tutor-feedback

fix(feedback-template-editor): improve feedback template sorting logic

coskun-kilinc committed:
fix(feedback-template-editor): add missing operator

Replaces a missing `**` operator in depth multiplier calculation.

coskun-kilinc committed:
refactor(feedback-template-editor): remove console logs

- Removed unnecessary console logs used for debugging sorted templates.
- Minor cleanup of redundant variable assignments.

satikaj committed:
Merge pull request #3 from coskun-kilinc/add-tutor-feedback

Add tutor feedback

satikaj committed:
fix: show description for connected outcome selection

satikaj committed:
fix: use task status color for recommended button

satikaj committed:
feat: add greeting and summary chip

satikaj committed:
fix: exclude greeting and summary from selected

satikaj committed:
refactor: separate outcome and feedback template editor

satikaj committed:
fix: prevent saving empty items

macite committed:
chore: update entity service

Fixes issues with task definitions not getting their id
when created. Which caused upload of
task sheets to fail until reloaded.

macite committed:
fix: enhance rollover messages

Make sure roll over in a unit indicates it
copies to a new teaching period or custom
dates.

macite committed:
chore(release): 8.0.34

macite committed:
fix: correct ability to edit teaching periods

macite committed:
fix: ensure an old grade is present in portfolio assessment

Fixes an issue where this somehow becomes null or undefined.

macite committed:
fix: ensure overseer images can be edited

macite committed:
chore(release): 8.0.35

macite committed:
fix: overseer image admin in unit and task

Correct connections and concurrency bug in loading overseer image when unit loaded.

macite committed:
chore: add proxy for running in dev using compose

macite committed:
chore(release): 8.0.36

macite committed:
fix: ensure overseer download get resources zip

macite committed:
Merge branch '8.0.x' of https://github.com/macite/doubtfire-web into 8.0.x

macite committed:
chore(release): 8.0.37

Lolretrorat committed:
fix: correct a typo in /edit_profile

macite committed:
feat: add option for institution logo

Embedded top left in toolbar.

CaptainBoggle committed:
fix: correct string manipulation for task period formatting

macite committed:
fix: reply to position and wrapping

macite committed:
chore: remove unused old angular tools

macite committed:
Merge branch 'add-tutor-feedback' of https://github.com/satikaj/doubtfire-web into satikaj-add-tutor-feedback

b0ink committed:
fix: support windows build

b0ink committed:
fix: support hot reloading on windows

b0ink committed:
chore: update package-lock.json

b0ink committed:
fix: update packages

b0ink committed:
chore: save esbuild to devdependencies

b0ink committed:
fix: lock ng2-pdf-viewer to 10.2.2

macite committed:
Merge branch 'fix/windows-build' of https://github.com/b0ink/doubtfire-web into b0ink-fix/windows-build

macite committed:
Merge branch 'b0ink-fix/windows-build' into 8.0.x

macite committed:
Merge branch '8.0.x' into satikaj-add-tutor-feedback

macite committed:
Merge branch 'bugfix/fix_typo_in_edit_profile' of https://github.com/Lolretrorat/doubtfire-web into Lolretrorat-bugfix/fix_typo_in_edit_profile

macite committed:
Merge branch 'Lolretrorat-bugfix/fix_typo_in_edit_profile' into 8.0.x

macite committed:
Merge branch 'feat/persistent-pdf-zoom' of https://github.com/b0ink/doubtfire-web into b0ink-feat/persistent-pdf-zoom

macite committed:
Merge branch 'b0ink-feat/persistent-pdf-zoom' into 8.0.x

macite committed:
Merge branch 'feat/pdfviewer-custom-start-page-8.0.x' of https://github.com/b0ink/doubtfire-web into b0ink-feat/pdfviewer-custom-start-page-8.0.x

macite committed:
Merge branch 'b0ink-feat/pdfviewer-custom-start-page-8.0.x' into 8.0.x

macite committed:
Merge branch 'migrate/confirmation-modal' of https://github.com/b0ink/doubtfire-web into b0ink-migrate/confirmation-modal

macite committed:
Merge branch 'b0ink-migrate/confirmation-modal' into 8.0.x

macite committed:
Merge branch '8.0.x' into satikaj-add-tutor-feedback

macite committed:
Merge branch 'development' of https://github.com/CaptainBoggle/doubtfire-web into 8.0.x

macite committed:
Merge branch '8.0.x' into satikaj-add-tutor-feedback

macite committed:
chore: add arm64 nx linux

macite committed:
chore: remove arm specific package

macite committed:
chore: remove arm specific package

macite committed:
chore: progress switch to refresh token secure cookie

macite committed:
feat: rebuild auth with refresh token

macite committed:
refactor: adjust handling of remember me

b0ink committed:
Merge remote-tracking branch 'upstream/8.0.x' into migrate/comments-modal

Commit:
feat(comments): enhanced drafting for messages using localStorage

macite committed:
chore(release): 10.0.0-1

macite committed:
fix: correct handling of errors on auth

macite committed:
chore(release): 10.0.0-2

macite committed:
fix: correct auth callback

macite committed:
chore(release): 10.0.0-3

Commit:
feat(comments): enhanced drafting for messages using localStorage

macite committed:
fix: ensure sign in waits for global load

macite committed:
chore(release): 10.0.0-4

ibi420 reviewed (approved):
Hello Jude, I have looked through your new feature, from my testing it runs smoothly and there are no errors. Thank you for the opportunity to review your work.

theiris6 reviewed (approved):
Hey Jude,
I’ve spent some time testing out the new feature you built, and everything’s working perfectly on my end. I didn’t run into any bugs or issues—it’s running really well. Thanks a lot for letting me take a look at it, appreciate being part of the review process!

nodogx commented:
Thank you so muuch @ibi420 and @theiris6 for getting this reviewed for me :)

macite committed:
chore: minor comment fix

macite committed:
fix: delay authorisation code until after authentication

Issue due to delay when loading from cookie.

b0ink commented:
Great work @nodogx 

One little thing i noticed is that if the draft comment is deleted without ever being sent, the key is still saved in local storage, perhaps the key could be deleted as soon as the value is empty again?

Other than that everything worked great on my end, tested it on windows chrome.

Commit:
fix(task-comment-composer): deleting the key when text is empty

nodogx committed:
Update col-xl.scss - Edited by mistake

nodogx committed:
Update dropdown-selector.scss - changed by mistake, changing back to original

nodogx commented:
@b0ink Thank for noticing this. I have fixed this, carefully tested this to ensure it works as intended. 

Now when the text is empty, it removes the key on the storage section.

b0ink commented:
Hi @nodogx 

When the input is empty, the storage key is removed, however, when i enter text into the input, then click on another student task submission, the key is also removed and the draft is lost - this may be a conflict in how the input is cleared when switching to a different submission

b0ink commented on src/app/tasks/task-comment-composer/task-comment-composer.component.ts:
```diff
@@ -94,8 +120,238 @@ export class TaskCommentComposerComponent implements DoCheck {
+    this.testLocalStorage();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.task && changes.task.currentValue !== changes.task.previousValue) {
+      const newTask = changes.task.currentValue;
+      const previousTask = changes.task.previousValue;
+
+      if (previousTask && !this.hasSubmittedComment) {
+        this.saveDraftForTask(previousTask);
+      }
```
the issue is most likely in your ngOnChanges hook here, i think this executes when the comment box has already been replaced and cleared for the new task submission, so it tries saving the draft with now empty content

you could probably just remove the save method entirely from ngOnChanges, you're already saving the draft to local storage on every keystroke, so it's a little redundant and you won't be losing any comment data

macite committed:
fix: correct broken transition hooks

macite committed:
fix: ensure transition in student pages update pdf view

nodogx committed:
Update task-comment-composer.component.ts

Removed redundant code which causes errors.

NgOnChanges now only looks to see if the task is changed. If the task is changed, it loads its content based on its key. Every keystroke is saved in the local storage; therefore, we draft is loaded, we can pull this from the storage straight away.

When its empty, it already deletes the whole key, so the content is no longer loaded.

nodogx commented:
NgOnChanges now only looks to see if the task is changed. If the task is changed, it loads its content based on its key. Every keystroke is saved in the local storage; therefore, we draft is loaded, we can pull this from the storage straight away.

When its empty, it already deletes the whole key, so the content is no longer loaded.

b0ink commented:
Looks good to me @nodogx

All working now, great job

macite committed:
fix: simplify setting of auto redirect on login

macite committed:
chore(release): 10.0.0-5

macite committed:
fix: only get task feedback templates if staff

aNebula reviewed (approved):
@nodogx LGTM.
Please open the upstream pull request with these changes against [doubtfire-lms/doubtfire-web 9.x branch](https://github.com/doubtfire-lms/doubtfire-web/tree/9.x)

macite committed:
fix: remove unnecessary safe pipe

macite committed:
feat: add spec con and flexible dates

macite committed:
chore(release): 10.0.0-6

nodogx commented:
> @nodogx LGTM. Please open the upstream pull request with these changes against [doubtfire-lms/doubtfire-web 9.x branch](https://github.com/doubtfire-lms/doubtfire-web/tree/9.x)

Thank you, this has been done! @aNebula

b0ink committed:
fix: welcome page redirect issues

macite committed:
Merge pull request #950 from b0ink/fix/auth-redirect

fix: welcome page redirect

b0ink committed:
Merge branch 'JackSCarroll:feature/add-jplag-frontend' into new/jplag-web

b0ink committed:
refactor: update jplag methods

b0ink committed:
chore: format and missing imports

b0ink committed:
build: init local jplag report viewer

b0ink committed:
feat: load jplag report viewer into iframe

b0ink committed:
chore: add back moss enum

b0ink committed:
chore: update jplag report icon

b0ink committed:
chore: remove unused code

b0ink committed:
chore: rename jplag methods

b0ink committed:
feat: native jplag report viewer state

b0ink committed:
chore: revert similarity type checks

b0ink committed:
fix: improve task similarity notice ui

b0ink committed:
build: update submodule url

b0ink committed:
feat: task and unit ilos ui

b0ink committed:
fix: enable unit name text wrapping

b0ink committed:
refactor: retrieve other student from similarity to open jplag comparison

b0ink committed:
build: ignore dev files from jplag build

b0ink committed:
chore: update jplag submodule

b0ink committed:
fix: ensure jplag iframe extends to full height

b0ink committed:
chore: move unit ilos below target grade selection

macite committed:
Merge pull request #951 from b0ink/new/jplag-web

New/jplag web

macite committed:
Merge pull request #953 from b0ink/fix/long-unit-name-styling

Fix: enable unit name text wrapping

macite committed:
Merge branch '10.0.x' into feat/ilos-ui

macite committed:
Merge pull request #952 from b0ink/feat/ilos-ui

Feat: task and unit ilos ui

macite committed:
Merge branch '10.0.x' into migrate/comments-modal

macite committed:
Merge pull request #869 from b0ink/migrate/comments-modal

Migrate/comments-modal component

macite committed:
Merge branch '10.0.x' into draft-comments
